third_party/move/move-model/src/builder/binary_module_loader.rs

```diff
@@ -0,0 +1,629 @@
+    }
+
+    fn type_param(&self, params: Option<&Vec<SourceName>>, i: usize) -> (Symbol, Loc) {
+        let (name, loc) = params
+            .and_then(|p| p.get(i))
+            .map(|(name, loc)| (name.to_owned(), Some(*loc)))
+            .unwrap_or_else(|| (format!("T{}", i), None));
+        (self.sym(&name), self.loc(loc))
+    }
+
+    fn loc(&self, loc: Option<MoveIrLoc>) -> Loc {
+        loc.map(|l| self.env.to_loc(&l))
+            .unwrap_or_else(|| self.module_loc.clone())
+    }
+
+    fn sym(&self, id: &str) -> Symbol {
+        self.env.symbol_pool().make(id)
+    }
+
+    fn error(&self, msg: impl AsRef<str>) {
+        self.env.error(
+            &self.module_loc,
+            &format!(
+                "while loading binary module `{}`: {}",
+                self.module_name.display(self.env),
+                msg.as_ref()
+            ),
+        )
+    }
+}
+
+fn make_module_name(module: language_storage::ModuleId, spool: &SymbolPool) -> ModuleName {
+    ModuleName::new(
+        Address::Numerical(module.address),
+        spool.make(module.name.as_str()),
+    )
+}
+
+fn type_params_logical_equal(params1: &[TypeParameter], params2: &[TypeParameter]) -> bool {
+    params1.len() == params2.len() && params1.iter().zip(params2).all(|(p1, p2)| p1.1 == p2.1)
+}
+
+fn params_logical_equal(params1: &[Parameter], params2: &[Parameter]) -> bool {
+    params1.len() == params2.len() && params1.iter().zip(params2).all(|(p1, p2)| p1.1 == p2.1)
+}
+
+/// Add well-known functions for vectors to module data. Those functions are instructions in
+/// the bytecode and do not have associated handles, but appear as regular functions in
+/// stackless bytecode and source.
+#[allow(non_snake_case)]
+fn add_well_known_vector_funs(env: &GlobalEnv, module_data: &mut ModuleData) {
+    // Vector module, add functions which are represented as bytecode
+    let t_params = vec![TypeParameter(
+        env.symbol_pool().make("T"),
+        TypeParameterKind::new(AbilitySet::EMPTY),
+        env.unknown_loc(),
+    )];
+    let t_T = Type::TypeParameter(0);
+    let t_ref_T = Type::Reference(ReferenceKind::Immutable, Box::new(t_T.clone()));
+    let t_mut_ref_T = Type::Reference(ReferenceKind::Mutable, Box::new(t_T.clone()));
+    let t_vec = Type::Vector(Box::new(Type::TypeParameter(0)));
+    let t_ref_vec = Type::Reference(ReferenceKind::Immutable, Box::new(t_vec.clone()));
+    let t_mut_ref_vec = Type::Reference(ReferenceKind::Mutable, Box::new(t_vec.clone()));
+    let t_u64 = Type::new_prim(PrimitiveType::U64);
+    let mk_param = |n, t| Parameter(env.symbol_pool().make(n), t, env.unknown_loc());
+    add_fun(
+        env,
+        module_data,
+        "empty",
+        t_params.clone(),
+        vec![],
+        t_vec.clone(),
+    );
+    add_fun(
+        env,
+        module_data,
+        "length",
+        t_params.clone(),
+        vec![mk_param("v", t_ref_vec.clone())],
```

Comment:
Not sure what the repercussions are for decompiling (probably none), but should these param names be `self` instead of `v` (and similarly below)?